src/storage/src/source/mod.rs

```diff
@@ -806,169 +866,640 @@ where
+                        // Not sure that's smart.
+                        Some(timestamp_frequency)
+                    }
+                    // TODO: Report these errors to the Healthchecker!
+                    Err(e) => {
+                        non_definite_errors.push(e);
+                        None
+                    }
+                    Ok(NextMessage::Finished) => {
+                        // This source reader is done. Yield one final
+                        // update of the source_upper.
+                        let unconsumed_partitions = source_reader.unconsumed_partitions();
+                        yield Some((
+                            std::mem::take(&mut untimestamped_messages),
+                            non_definite_errors.drain(..).collect_vec(),
+                            unconsumed_partitions,
+                            source_upper.clone(),
+                        ));
+
+                        // Then, let the consumer know we're done.
+                        yield None;
+
+                        return;
+                    }
+                };
 
-                        timestamper.advance().await;
-
-                        // TODO(petrosagg): compaction should be driven by AllowCompaction commands
-                        // coming from the storage controller
-                        let new_ts_upper = timestamper
-                            .reclock_frontier(&source_upper)
-                            .expect("compacted past upper");
-                        if PartialOrder::less_than(&ts_upper, &new_ts_upper) {
-                            // Downgrade our capability to the new upper, but only if has changed
-                            // since last time we downgraded it. Note property #2 documented above
-                            // is critical here: once this capability is downgraded, that timestamp
-                            // is sealed forever.
-                            ts_upper = new_ts_upper.clone();
-                            yield Event::Progress(new_ts_upper);
-                        }
-                        if source_stream.is_done() {
-                            yield Event::Progress(Antichain::new());
-                            break;
-                        }
+                // TODO: We could fetch more than 1 message before doing the
+                // tick check because doing this logic per consumed message
+                // seems somewhat costly.
+
+                let now = Instant::now();
+
+                if let Some(sleep_time) = sleep_time {
+                    // Sleep the min of the suggested sleep time and what time
+                    // remains until the next tick should happen.
+                    let sleep_time =
+                        std::cmp::min(sleep_time, tick_interval + (last_tick_time - now));
+
+                    tokio::time::sleep(sleep_time).await;
```

Comment:
you may prefer handling this with a tokio `Interval` with a chosen missed tick behavior: https://docs.rs/tokio/latest/tokio/time/struct.Interval.html#method.set_missed_tick_behavior, to simplify some of this manual logic handling

Reply:
Can this be done without a `select!`? I thought it could not, and most of the complexity in this part comes from getting rid of that `select!`. This is in flux, Petros also has thoughts about this that he mentioned to me on slack.

Reply:
it should be able to be awaited like the `sleep` here, `interval.tick()`